### PR TITLE
rm:linux: Add how to build with the clang toolchain

### PR DESCRIPTION
**Pull request by quaresmajose**
Signed-off-by: Jose Quaresma <jose.quaresma@foundries.io>

## Readiness

* [x] Merge (pending reviews)
* [ ] Merge after _date or event_
* [ ] Draft

## Overview

This merge add some notes on who to use the clang toolchain.

## Checklist

_Optional. Add a 'x' to steps taken._
_You can fill this out after opening the PR. "Did I..."_

* [ ] Run spelling and grammar check, preferably with linter.
* [ ] Avoid changing any header associated with a link/reference.
* [ ] Step through instructions (or ask someone to do so).
* [ ] Review for [wordiness](https://languagetool.org/insights/post/wordiness/)
* [ ] Match tone and style of page/section.
* [x] Run `make linkcheck`.
* [x] View HTML in a browser to check rendering.
* [ ] Use [semantic newlines](https://bobheadxi.dev/semantic-line-breaks/).
* [ ] follow best practices for commits.
  * [ ] Descriptive title written in the imperative.
  * [ ] Include brief overview of QA steps taken.
  * [ ] Mention any related issues numbers.
  * [x] End message with sign off/DCO line (`-s, --signoff`).
  * [ ] Sign commit with your gpg key (`-S, --gpg-sign`).
  * [ ] Squash commits if needed.
* [x] Request PR review by a technical writer and at least one peer.

## Comments

_Any thing else that a maintainer/reviewer should know._
_This could include potential issues, rational for approach, etc._

### CODE REVIEW

File: source/reference-manual/linux/toolchain.rst [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+Exists however some issues when using the ``clang`` and ``llvm`` as default,
+because some recipes are refusing to be build with this configuration. To adresss this issues
+is recomendend to add a new file where we will add the tweeks needed to build
+sucessfully. In the public LmP we use the ``meta-lmp/meta-lmp-base/conf/distro/include/non-clangable.inc``
+for that propuse.
+
+Adding the same on the ``meta-subscriber-overrides`` can be a solution
+where we collect all the ``clang`` tweeks needed. This configuration
+can be added in ``meta-subscriber-overrides/conf/layer.conf`` or somewere in another global
```
Comment (ricardosalveti) [inline source/reference-manual/linux/toolchain.rst:22]: We do use 'conf/machine/include/lmp-factory-custom.inc' for global settings that should be factory specific, maybe use that instead (needs testing due include ordering).

File: source/reference-manual/linux/toolchain.rst [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+.. prompt:: text
+
+  require meta-subscriber-overrides/conf/non-clangable.inc
+
+
+Changing the toolchain
+----------------------
+
+To change the default values disabling the ``clang`` compiler globaly and use the ``gcc``
+we need to change the ``TOOLCHAIN`` value in ``meta-subscriber-overrides/conf/layer.conf``
```
Comment (quaresmajose) [inline source/reference-manual/linux/toolchain.rst:34]: it can be better to place everything clang related inside the `non-clangable.inc` and it's easier to find ?

File: source/reference-manual/linux/toolchain.rst (hunk 1 above)
Comment (quaresmajose) [reply source/reference-manual/linux/toolchain.rst:22]: `conf/machine/include/lmp-factory-custom.inc` looks better maybe the non-clangable.inc can be placed on the same folder `conf/machine/include/non-clangable.inc`

File: source/reference-manual/linux/toolchain.rst [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+because some recipes are refusing to be build with this configuration. To adresss this issues
+is recomendend to add a new file where we will add the tweeks needed to build
+sucessfully. In the public LmP layer ``meta-lmp`` we use the
+``meta-lmp-base/conf/distro/include/non-clangable.inc``
+for that propuse.
+
+Adding the same on the factory layer ``meta-subscriber-overrides`` can be a solution
+where we will collect all the ``clang`` tweeks needed. This configuration
+can be added in ``conf/machine/include/lmp-factory-custom.inc``
+or somewere in another global configurations file.
```
Comment (angolini) [inline source/reference-manual/linux/toolchain.rst:24]: This piece is confusing to me. I mean, it's unclear whether I should create both files or choose one. And in case I must choose, how do I choose?

I would rephrase it by starting by g pointing to the `meta-lmp-base/conf/distro/include/non-clangable.inc` file. Some thing as: "There is a file in LmP with some changes, but it's recommended to create another one <here>". Or only mention the `conf/machine/include/lmp-factory-custom.inc` and nothing more.

The user with experience will know how to organize his source code the way he wants. We can point the non-experienced user to how to do it.

File: source/reference-manual/linux/toolchain.rst [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+.. prompt:: text
+
+  require conf/machine/include/non-clangable.inc
+
+So we will use the factory file ``conf/machine/include/non-clangable.inc``
```
Comment (angolini) [inline source/reference-manual/linux/toolchain.rst:30]: Another way to say all the possible files we can change is by making a bullet list. But then I would be wondering how to choose from the list.

File: source/reference-manual/linux/toolchain.rst [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+Changing the toolchain
+----------------------
+
+To change the default values disabling the ``clang`` compiler globaly and use the ``gcc``
+we need to change the ``TOOLCHAIN`` value.
+
+.. prompt:: text
+
+  TOOLCHAIN = "gcc"
```
Comment (angolini) [inline source/reference-manual/linux/toolchain.rst:42]: where?

File: source/reference-manual/linux/toolchain.rst [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+Changing the compiler per recipe is also possible and it can be done using a bbappend or
```
Comment (angolini) [inline source/reference-manual/linux/toolchain.rst:44]:
```suggestion
Changing the compiler per recipe is also possible, and it can be done using a `.bbappend` or
```

File: source/reference-manual/linux/toolchain.rst [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+with ``TOOLCHAIN = "gcc"`` or in the global factory ``non-clangable.inc``
```
Comment (angolini) [inline source/reference-manual/linux/toolchain.rst:45]:
```suggestion
with ``TOOLCHAIN = "gcc"`` or in the factory, changing the file ``conf/machine/include/non-clangable.inc``, from ``meta-subscriber-overrides``
```

File: source/reference-manual/linux/toolchain.rst [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+  TOOLCHAIN:pn-<recipe> = "gcc"
+
+
+Costumize the default toolchain
+-------------------------------
+
+When using the clang toolchain we have the ``toolchain-clang`` override that can be used
+to do the costumization and this will only take effect when clang is in use.
+All of this costumizations can be placed on the factory ``non-clangable.inc``
```
Comment (angolini) [inline source/reference-manual/linux/toolchain.rst:57]:
```suggestion
All of these customizations can be placed in the factory by changing the file ``conf/machine/include/non-clangable.inc``, from ``meta-subscriber-overrides``.
```

File: source/reference-manual/linux/toolchain.rst [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+.. prompt:: text
+
+  STRIP:pn-<recipe>:toolchain-clang = "${HOST_PREFIX}strip"
+  OBJCOPY:pn-<recipe>:toolchain-clang = "${HOST_PREFIX}objcopy"
+  COMPILER_RT:pn-<recipe>:toolchain-clang = "-rtlib=libgcc --unwindlib=libgcc"
+
+
+Some more infiornation can be found on https://github.com/kraj/meta-clang
```
Comment (angolini) [inline source/reference-manual/linux/toolchain.rst:66]:
```suggestion
Some more information can be found on https://github.com/kraj/meta-clang
```

File: source/reference-manual/linux/toolchain.rst [hunk 10]
```diff
@@ -0,0 +1,66 @@
+.. _ref-toolchain:
+
+Toolchain
+=========
+
```
Comment (angolini) [inline source/reference-manual/linux/toolchain.rst:5]: Perhaps start by saying that there are 2 available toolchains: GCC and clang.

**Review by angolini — APPROVED** [review]
LGTM, only minor comments

File: source/reference-manual/linux/toolchain.rst (hunk 3 above)
Comment (quaresmajose) [reply source/reference-manual/linux/toolchain.rst]: I tried to rewrite this in another way

**Comment by doanac** [comment]
Docs for 81d6d1a8feae5216c8100745dda21cc395e57d8c are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1864/docs/artifacts/html/index.html

**Comment by quaresmajose** [comment]
@angolini add all your suggestions, thanks for the review.

**Comment by gcgrey** [comment]
Jose, good work getting this written up. 
Copying Katrina on this as it needs a pass through for spelling and grammar. 

George

**Comment by quaresmajose** [comment]
Thnaks @gcgrey. Have pinged the Katrina

File: source/reference-manual/linux/toolchain.rst [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+tweeks will be made to build sucessfully.
+
+In the public LmP layer ``meta-lmp`` we use the
+``meta-lmp-base/conf/distro/include/non-clangable.inc``
+for that propuse.
+Adding another ``non-clangable.inc`` on the factory layer ``meta-subscriber-overrides``
+can be a solution, where we will collect all the ``clang`` tweeks needed.
+
+This configuration can be added in the factory global configurations file
+``conf/machine/include/lmp-factory-custom.inc``
```
Comment (kprosise) [inline source/reference-manual/linux/toolchain.rst:27]:
```suggestion
``conf/machine/include/lmp-factory-custom.inc``:
```

File: source/reference-manual/linux/toolchain.rst [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+To address these issues is recomendend to add a new file where these 
+tweeks will be made to build sucessfully.
+
+In the public LmP layer ``meta-lmp`` we use the
+``meta-lmp-base/conf/distro/include/non-clangable.inc``
+for that propuse.
+Adding another ``non-clangable.inc`` on the factory layer ``meta-subscriber-overrides``
+can be a solution, where we will collect all the ``clang`` tweeks needed.
+
+This configuration can be added in the factory global configurations file
```
Comment (kprosise) [inline source/reference-manual/linux/toolchain.rst:26]:
```suggestion
This can be added in the factory global configuration file
```

File: source/reference-manual/linux/toolchain.rst [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+``conf/machine/include/lmp-factory-custom.inc``
+
+.. prompt:: text
+
+  require conf/machine/include/non-clangable.inc
+
+So we will use the factory file ``conf/machine/include/non-clangable.inc``
+to customize everything ``clang`` related.
+
+Changing the toolchain
```
Comment (kprosise) [inline source/reference-manual/linux/toolchain.rst:36]:
```suggestion
Changing the Toolchain
```

File: source/reference-manual/linux/toolchain.rst [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+----------------------
+
+To change the default values disabling the ``clang`` compiler globaly and use the ``gcc``
```
Comment (kprosise) [inline source/reference-manual/linux/toolchain.rst:39]:
```suggestion
To change the default values disabling ``clang`` globally so as to use  ``gcc``,
```

File: source/reference-manual/linux/toolchain.rst [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+can be a solution, where we will collect all the ``clang`` tweeks needed.
+
+This configuration can be added in the factory global configurations file
+``conf/machine/include/lmp-factory-custom.inc``
+
+.. prompt:: text
+
+  require conf/machine/include/non-clangable.inc
+
+So we will use the factory file ``conf/machine/include/non-clangable.inc``
```
Comment (kprosise) [inline source/reference-manual/linux/toolchain.rst:33]:
```suggestion
**So we will use the factory file ``conf/machine/include/non-clangable.inc``**
```

File: source/reference-manual/linux/toolchain.rst [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+----------------------
+
+To change the default values disabling the ``clang`` compiler globaly and use the ``gcc``
+we need to change the ``TOOLCHAIN`` value.
+
+.. prompt:: text
+
+  TOOLCHAIN = "gcc"
+
+Changing the compiler per recipe is also possible and to do that we add the
```
Comment (kprosise) [inline source/reference-manual/linux/toolchain.rst:46]:
```suggestion
Changing the compiler per recipe is also possible. To do that we add the
```

File: source/reference-manual/linux/toolchain.rst [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+line above in a `.bbappend` for the recipe.
+This can also be done in the factory, changing the file
+``conf/machine/include/non-clangable.inc``, from ``meta-subscriber-overrides``.
+
+.. prompt:: text
+
+  TOOLCHAIN:pn-<recipe> = "gcc"
+
+
+Costumize the default toolchain
```
Comment (kprosise) [inline source/reference-manual/linux/toolchain.rst:56]:
```suggestion
Customizing the Default Toolchain
```

File: source/reference-manual/linux/toolchain.rst [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+-------------------------------
+
+When using the clang toolchain we have the ``toolchain-clang`` override that can be used
+to do the costumization and this will only take effect when clang is in use.
```
Comment (kprosise) [inline source/reference-manual/linux/toolchain.rst:60]:
```suggestion
to do the customization and this will only take effect when clang is in use.
```

File: source/reference-manual/linux/toolchain.rst [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+``conf/machine/include/non-clangable.inc``, from ``meta-subscriber-overrides``.
+
+.. prompt:: text
+
+  STRIP:pn-<recipe>:toolchain-clang = "${HOST_PREFIX}strip"
+  OBJCOPY:pn-<recipe>:toolchain-clang = "${HOST_PREFIX}objcopy"
+  COMPILER_RT:pn-<recipe>:toolchain-clang = "-rtlib=libgcc --unwindlib=libgcc"
+
+
+Some more information can be found on https://github.com/kraj/meta-clang
```
Comment (kprosise) [inline source/reference-manual/linux/toolchain.rst:71]:
```suggestion
More information can be found via the `meta-clang repository <https://github.com/kraj/meta-clang>`_.
```

**Review by kprosise — COMMENTED** [review]
Handful of suggestions, focusing primarily on spelling and grammar.

**Comment by quaresmajose** [comment]
I addressees all of your comment.
Thank you
